CURA-6387 Add HttpRequestManager

LipuFei opened the pull request.

fieldOfView commented:
This is probably not in the design spec, but it would almost be s shame not to (optionally) handle timeouts in this requestmanager.

LipuFei commented:
@fieldOfView I will take a look later. This ticket is not in priority and not planned, so right now this PR is merely a cleanup of the WIP branch I did earlier.

LipuFei commented:
@fieldOfView timeout has been added in a8345ad. I marked it as "experimental" because it can cause problems in unstable networks.

LipuFei commented:
FYI, I tested the timeout with Cura Toolbox by manually changing the download URL to a very large file, for example http://ftp2.nluug.nl/languages/qt/archive/qt/5.14/5.14.0/single/qt-everywhere-src-5.14.0.tar.xz (522MB), and specify a timeout of 5 seconds or so. I checked the log to see if things were happening in the correct order.

fieldOfView commented:
I see the timeout mainly for use with things like status requests to the UM3 networking API (and, as you might have guessed, OctoPrint). Things that should be almost instantaneous but break when eg the network goes away. For that usecase the suggested timeout implementation should work great.

For longer transfers, if you could add functionality to "reset" the timeout, that could be called whenever there is upload/download progress.

LipuFei commented on UM/TaskManagement/HttpRequestManager.py:
```diff
@@ -0,0 +1,457 @@
+
+        # Make sure that Qt handles redirects
+        if hasattr(QNetworkRequest, "FollowRedirectsAttribute"):
+            # Patch for Qt 5.6-5.8
```
This was there so I kept it.

ninovanhooff replied:
Hmm. The comment suggests that this has been fixed in 5.9, which means we don't need it.

LipuFei commented on UM/TaskManagement/HttpRequestManager.py:
```diff
@@ -0,0 +1,457 @@
+
+        Logger.log("d", "%s finished", request_data)
+        with self._request_lock:
+            # safeguard: ake sure that we have the reply in the currently in-progress requests set.
```
Fixed in 35a1a570

LipuFei commented on UM/TaskManagement/HttpRequestManager.py:
```diff
@@ -0,0 +1,457 @@
+                       upload_progress_callback: Optional[Callable[[int, int], None]] = None,
+                       timeout: Optional[float] = None) -> "HttpRequestData":
+        assert(timeout is None or timeout > 0)
+        if timeout is not None and timeout <= 0:
```
Although we don't do it, if we run Python in the optimized mode, which will generate `*.pyo` instead of `*.pyc`, certain commands will be ignored, such as `__debug__` and `assert()`. Basically I added the `if`-statement for the normal case, but we can remove the `assert`.

LipuFei commented on UM/TaskManagement/HttpRequestManager.py:
```diff
@@ -0,0 +1,457 @@
+
+    # Processes the next request in the pending queue. Stops if there is no more pending requests. It also stops if
+    # the maximum number of concurrent requests has been reached.
+    def _processRequestsInQueue(self) -> None:
```
Renamed in c903e4cad1859a6fa21405caa37f69b48db0e1b5

LipuFei replied:
Missed a few. Fixed in e92959d2

LipuFei commented on UM/TaskManagement/HttpRequestManager.py:
```diff
@@ -0,0 +1,457 @@
+            self._application.callLater(request_data.callback, request_data.reply)
+
+        # continue to process the next request
+        self._processRequestsInQueue()
```
I extended the doc for 42b3f5172ddd745f92a06b9feddca5064fae0f3b. If you think it needs more explanations, feel free to point out.

ninovanhooff replied:
Ah right, so the queue contains requests which are not sent yet because there are already `max_concurrent_requests` in progress?. Fine.

LipuFei commented:
> I think this is a great improvement. Good comments and clean code. Well done. I do agree with @fieldOfView that this "total request duration timeout" is unintuitive and you should reset the timeout on progress updates

I agree, but I will try to spend more time on this and hopefully make some test cases so we can be more certain that it's doing things correctly.

LipuFei commented:
My main concern about the "proper" timeout implementation is the CPU usage and the complexity. Right now, this network manager will only create signal connections if download/upload callbacks need to be handled by the using component, and most using components do not handle it. If we do a timeout reset + recalculation for every download/upload event, it will cost a lot more resources. It's a pity that Qt doesn't seem to support timeout at the moment (https://doc.qt.io/qt-5/qnetworkrequest.html). IMO, the solutions we have been discussing so far are still workarounds.

LipuFei commented:
@fieldOfView @ninovanhooff  I've done some rework:
 - Reimplemented the timeout to be the max wait time between each response we get from the server.
 - Added a `TaskManager` class for easier callback scheduling and remove the dependency on `Application` to do any callback scheduling.
 - Added unit tests to test (1) a successful download, (2) a 404 error, and (3) timeout handling.
 - Updated docs.
